lib/options.nix

```diff
@@ -368,4 +369,63 @@ rec {
         else
           example;
     };
+
+  mkLazyLoadOption =
+    {
+      originalName ? "this plugin",
+      cfg ? { },
+      lazyLoad ? { },
+      optionsForPlugin ? false,
+    }:
+    let
+      pluginDefault = {
+        enable = false;
+        name = originalName;
+      } // lazyLoad;
+    in
+    mkOption {
+      description = "Lazy-load settings for ${originalName}.";
+      type =
+        with nixvimTypes;
+        submodule {
+          options = with defaultNullOpts; {
+            # enable option only has an effect when it is a sub-option for a plugin generated by mkNeovimPlugin
+            enable = mkOption {
+              type = bool;
+              default = pluginDefault.enable;
+              description = "Enable lazy-loading for ${originalName}";
+              # Only show this option when it is a sub-option for a plugin
+              visible = optionsForPlugin;
+            };
+            name = mkOption {
+              type = str;
+              default = pluginDefault.name;
+              description = "The plugin's name (not the module name). This is what is passed to the load(name) function.";
+            };
+            enabledInSpec = mkStrLuaFnOr bool pluginDefault.enabledInSpec or null ''
+              When false, or if the function returns false, then ${originalName} will not be included in the spec.
+              This option corresponds to the `enabled` property of lz.n.
+            '';
+            beforeAll = mkLuaFn pluginDefault.beforeAll
+              or null "Always executed before any plugins are loaded.";
+            before = mkLuaFn pluginDefault.before or null "Executed before ${originalName} is loaded.";
+            after = mkLuaFn pluginDefault.after or null "Executed after ${originalName} is loaded.";
+            event =
+              mkNullable (listOf str) pluginDefault.event or null
+                "Lazy-load on event. Events can be specified as `BufEnter` or with a pattern like `BufEnter *.lua`";
+            cmd = mkNullable (listOf str) pluginDefault.cmd or null "Lazy-load on command.";
+            ft = mkNullable (listOf str) pluginDefault.ft or null "Lazy-load on filetype.";
+            keys = mkNullable (listOf nixvimKeymaps.mapOptionSubmodule) pluginDefault.keys
+              or null "Lazy-load on key mapping. Use the same format as `config.keymaps`.";
+            colorscheme = mkNullable (listOf str) pluginDefault.colorscheme or null "Lazy-load on colorscheme.";
+            priority = mkNullable number pluginDefault.priority or null ''
+              Only useful for start plugins (not lazy-loaded) to force loading certain plugins first. 
+              Default priority is 50 (or 1000 if colorscheme is set).
+            '';
+            load = mkLuaFn pluginDefault.load
+              or null "Can be used to override the vim.g.lz_n.load() function for ${originalName}.";
+          };
```

Comment:
How closely tied are these options to lz.n? Ideally the per-plugin options would be common to both lz.n, lazy.nvim and any other lazyload implementation we end up supporting in the future.

This could mean not making all options available per-plugin (users could still define them via the top-level option) and/or adding warnings/documentation indicating that some options only apply to specific implementation(s).

Reply:
> How closely tied are these options to lz.n? Ideally the per-plugin options would be common to both lz.n, lazy.nvim and any other lazyload implementation we end up supporting in the future.

Most options are shared by both lz.n and lazy.nvim, just with different names that would have to be adapted when setting them for each backend.
The differences are:
- `before`: is only defined by lz.n. lazy.nvim only allows `beforeAll`.
- `colorscheme`: lazy.nvim lazy-loads colorschemes automatically, so it wouldn't be set for it.
- `load`: This is specific for lz.n.




> This could mean not making all options available per-plugin (users could still define them via the top-level option) and/or adding warnings/documentation indicating that some options only apply to specific implementation(s).

What do you mean by "define them via the top-level option"?

Another idea I had is to have the subset of optrons that apply to all implementations at top-level, and a FreeForm option that is passed to the specific implementation being used. Just like `plugins.*.settings` is passed to `require(*).setup`.
What do you think fits best?

Reply:
> What do you mean by "define them via the top-level option"?

I mean that if `plugins.*.lazyload` doesn't provide the option a user needs, they can directly use `plugins.lazy` or `plugins.lz-n`.

> Most options are shared by both lz.n and lazy.nvim, just with different names that would have to be adapted when setting them for each backend.

My view here is that we should define our own options here, without trying to match lazy.nvim or lz.n _exactly_. We can then map whatever options we decided on to the implementing plugin's settings internally.

This should keep our per-plugin option de-coupled from the implementing plugins' options, allowing for future changes both on our end and in upstream plugins.

Reply:
> Another idea I had is to have the subset of optrons that apply to all implementations at top-level, and a FreeForm option that is passed to the specific implementation being used. Just like `plugins.*.settings` is passed to `require(*).setup`.
> What do you think fits best?

I'm not sure I follow 100%, but freeform and rfc42 are usually the way to go when dealing directly with a plugin's options.